2021-12-27-getting-started-with-purenix.md

```diff
@@ -0,0 +1,217 @@
+
+One of the main build tools for PureScript is
+[Spago](https://github.com/purescript/spago).  Spago is somewhat similar to
+Haskell's `stack` or `cabal`, Rust's `cargo`, etc.
+
+One interesting feature of the PureScript compiler and Spago is that there is
+good support for switching out the backend code generator.  The code generator
+built into the main PureScript compiler is for JavaScript, but there are
+alternative code generators for other languages.  PureNix is an alternative
+backend code generator for Nix.
+
+## PureNix
+
+PureNix works together with PureScript and Spago to compile PureScript code
+into Nix code.  Here's a small example of PureScript code, and the Nix code it
+gets compiled to:
+
+```purescript
+module Main where
+
+greeting :: String
+greeting = "Hello, world!"
+
+identity :: forall a. a -> a
+identity a = a
+
+data Maybe a = Nothing | Just a
+
+fromMaybe :: forall a. a -> Maybe a -> a
+fromMaybe a Nothing = a
+fromMaybe _ (Just a) = a
+```
+
+PureNix compiles this PureScript code to the following Nix code[^1]:
+
+```nix
+let
+  greeting = "Hello, world!";
+  Nothing = {__tag = "Nothing";};
+  Just = value0:
+    { __tag = "Just";
+      __field0 = value0;
+    };
+  fromMaybe = v: v1:
+    let
+      __pattern0 = __fail: if v1.__tag == "Nothing" then let a = v; in a else __fail;
+      __pattern1 = __fail: if v1.__tag == "Just" then let a = v1.__field0; in a else __fail;
+      __patternFail = builtins.throw "Pattern match failure in src/Main.purs at 11:1 - 11:41";
+    in
+      __pattern0 (__pattern1 __patternFail);
+  identity = v: v;
+in
+  {inherit greeting Nothing Just fromMaybe;}
+```
+
+Except for the pattern matching in `fromMaybe`, this should look pretty
+straight-forward.  You can see the PureScript string `greeting` becomes a
+simple Nix string.  `identity` has become a Nix function.  The constructors for
+`Maybe` (`Nothing` and `Just`) have become Nix functions[^2].  `fromMaybe` has
+become a Nix function.  If you work through the `__patternX` continuations in
+`fromMaybe`, you can see how they
+[emulate the pattern matching](https://jonascarpay.com/posts/2021-11-08-nix-adts.html)
+in the PureScript `fromMaybe` function.
+
+Calling functions written in PureScript from Nix is quite fun.  It feels quite
+similar to working in the Haskell or PureScript REPL.  Here is an example of
+using the above functions from the Nix REPL:
+
+```console
+$ nix repl ./output/Main/default.nix
+nix-repl> identity greeting
+"Hello, world!"
+nix-repl> fromMaybe "bye" (Just "hello")
+"hello"
+nix-repl> fromMaybe "bye" Nothing
+"bye"
+```
+
+Now that you have a taste of PureNix, the next section explains how to get
+started actually writing PureNix code.
+
+## Getting Started
+
+There is a post on the NixOS Discourse that explains how to
+[put together a development environment for PureNix](https://discourse.nixos.org/t/purenix-nix-backend-for-purescript/15756/3).
```

Comment:
This may be worth inlining? Or making into a separate post in the series?

Reply:
Yeah, that's a good point.  I was basically doing the laziest thing I could here.

I thought about this a little bit, and I think what would be ideal is to have the content from https://discourse.nixos.org/t/purenix-nix-backend-for-purescript/15756/3 as an official document in the purenix repo.

I'll clean it up that post a bit and send it as a PR to the purenix repo.

Reply:
I've sent a PR for this as https://github.com/purenix-org/purenix/pull/38.

Reply:
Fixed in this blog post in 34c2a58.